tools/rpmuncompress.c

```diff
@@ -78,16 +84,81 @@ static char *doUncompress(const char *fn)
     return cmd;
 }
 
+/**
+ * Detect if an archive has a single top level entry, and it's a directory.
+ *
+ * @param path	path of the archive
+ * @return	1 if archive as only a directory as top level entry,
+ * 		0 if it contains multiple top level entries or a single file
+ * 		-1 on archive error
+ */
+static int singleRoot(const char *path)
+{
+	struct archive *a;
+	struct archive_entry *entry;
+	int r, ret, rootLen;
+	char *rootName;
+
+	a = archive_read_new();
+	archive_read_support_filter_all(a);
+	archive_read_support_format_all(a);
+	r = archive_read_open_filename(a, path, 10240);
+	if (r != ARCHIVE_OK) {
+	    ret = -1;
+	    goto afree;
+	}
+	if (archive_read_next_header(a, &entry) != ARCHIVE_OK) {
+	    ret = -1;
+	    goto afree;
+	}
+	rootName = strdup(archive_entry_pathname(entry));
```

Comment:
Rpm code should generally use xstrdup() aka rstrdup() instead to follow the central abort-on-failure mode. strdup() use should be limited to cases where you can meaningfully deal with an OOM (this is not one of them), and actually do so.